src/chem/modis_aod2ioda.py

```diff
@@ -0,0 +1,183 @@
+        self.filenames = filenames
+        self.writer = writer
+        self.varDict = defaultdict(lambda: defaultdict(dict))
+        self.outdata = defaultdict(lambda: DefaultOrderedDict(OrderedDict))
+        self.loc_mdata = defaultdict(lambda: DefaultOrderedDict(OrderedDict))
+        self.var_mdata = defaultdict(lambda: DefaultOrderedDict(OrderedDict))
+        self.units = {}
+        self._read()
+
+    def _read(self):
+        # set up variable names for IODA
+        for iodavar in ['Aerosol_optical_depth']:
+            self.varDict[iodavar]['valKey'] = iodavar, self.writer.OvalName()
+            self.varDict[iodavar]['errKey'] = iodavar, self.writer.OerrName()
+            self.varDict[iodavar]['qcKey'] = iodavar, self.writer.OqcName()
+
+        # Set up modis time
+
+        modis_time = {}
+
+        # loop through input filenames
+        first = True
+        for f in self.filenames:
+            ncd = nc.Dataset(f, 'r')
+            ncd_str = str(ncd)
+            if 'Terra' in ncd_str:
+                AttrData['platform'] = 'Terra'
+            elif 'Aqua' in ncd_str:
+                AttrData['platform'] = 'Aqua'
+            if 'MODIS' in ncd_str:
+                AttrData['sensor'] = 'MODIS'
+
+            #  Get variables
+            modis_time = ncd.variables['Scan_Start_Time'][:].ravel()
+
+            # convert time to date_string
+            lats = ncd.variables['Latitude'][:].ravel()
+            lons = ncd.variables['Longitude'][:].ravel()
+            aod = ncd.variables['AOD_550_Dark_Target_Deep_Blue_Combined'][:].ravel()
+            land_sea_flag = ncd.variables['Land_sea_Flag'][:].ravel()
+            QC_flag = ncd.variables['Land_Ocean_Quality_Flag'][:].ravel()
+            sol_zen = ncd.variables['Solar_Zenith'][:].ravel()
+            sol_zen = sol_zen
+            sen_zen = ncd.variables['Sensor_Zenith'][:].ravel()
+            sen_zen = sen_zen
+            unc_land = ncd.variables['Deep_Blue_Aerosol_Optical_Depth_550_Land_Estimated_Uncertainty'][:].ravel()
+            unc_land = unc_land
+
+            time_1 = timedelta(seconds=1)
+            obs_time = np.empty_like(QC_flag, dtype=object)
+            obs_time = datetime.fromisoformat('1993-01-01') + modis_time*time_1
+
+            # uncertainty estimates:
+            # From MODIS file (over ocean) and Levy, 2010 (over land)
+            # flag = 0 (ocean) 1(land) 2(coastal)
+
+            over_ocean = np.logical_not(land_sea_flag > 0)
+            over_land = np.logical_not(land_sea_flag == 0)
+            UNC = np.where(over_land, unc_land, np.add(0.05, np.multiply(0.15, aod)))
+            UNC_2 = UNC[0]
+
+            if first:
+                self.loc_mdata['latitude'] = lats
+                self.loc_mdata['longitude'] = lons
+                self.loc_mdata['datetime'] = self.writer.FillNcVector(obs_time, "datetime")
+                self.loc_mdata['latitude'] = np.concatenate((self.loc_mdata['latitude'], lats[0:1]))
```

Comment:
This is due to a problem that I have so far been unable to resolve. I could probably use some help debugging here. This is only needed for the ctest because the code fails when processing just one file. Generally many files will be processed at a time.  The error message  occurs at the "writer.BuildNetcdf" step, and only for "aod_class.outdata".  The message is "array type unknown >i2".    My guess is that it is related to the array format coming from the HDF4 file. I've tried changing the dtype to float32 (it came in as float64), but this had no impact.  The added value at the end of the arrays are "undefined" so they have no impact on the dataset, but I agree that it is technically incorrect.  Input from @srherbener would be helpful here.